### ISSUE TITLE
No hard-coding of grammars and databases in the Haskell source code

### ISSUE THREAD

**Issue by heatherleaf** [MUSTE-Project/MULLE#57]
The goal of MULLE is to have a tool that anyone can install without any knowledge of Haskell. 

So, the info in the following files should be moved to some kind of config files:

- `muste-ajax/app/DbInit/Data.hs`
- `muste/src/Muste/Grammar/Grammars.hs`

Additionally, it would be nice if the tests and benchmarks could load their tests from a config file.

**Comment by heatherleaf** [comment]
I like yaml myself. And it could look something like this:

```
grammars:
  Prima: novo_modo/Prima.pgf    # path to grammar file
  Secunda: novo_modo/Secunda.pgf
  Exemplum: exemplum/Exemplum.pgf

languages:
  ExemplumAra:
    direction: rtl  # default ltr
  ExemplumChi: 
    spacing: 50%  # default 100%

# lessons are ordered - they have to be shown in some order in the GUI
lessons:
- name: PrimaLatSwe
  description: Lesson 1 from Novo Modo
  grammar: Prima
  srcLang: PrimaLat
  trgLang: PrimaSwe
  # sizeLimit: 3  # default value: null
  depthLimit: 3  # default value: null
  # enabled: true  # default value
  repeatable: false  # default value: true
- name: ExemplumEngAra
  description: Example lesson English-Arabic
  grammar: Exemplum
  srcLang: ExemplumEng
  trgLang: ExemplumAra
  sizeLimit: 3

# exercises are based on basic lesson keys
PrimaLatSwe:
- ["vinum sapiens est", "han är vis"]
- ["Augustus imperium tenet" , "imperatorn håller riket"]
- - Augustus felix est
  - vännen är lycklig

ExemplumEngAra:
- - many kings love Paris
  -  يَقْرَأُ مَلِكٌ جوَيِّدٌ كِتابً 
- - ...
  - ...
```

**Comment by fredefox** [comment]
Yes Yaml is amazing. This shouldn't be too difficult to implement. `muste-ajax` is already using Yaml for some configuration options.

As for the pgf-grammars. I've also kinda changed my mind about how this is currently implemented. Ideally the application should load the compiled grammar files from a prioritized list of data-directory e.g. in the case of XDG compliant operating systems it would be something like:

    ./grammars
    $HOME/.local/share/muste/grammars/
    /usr/share/muste/grammars

**Comment by heatherleaf** [comment]
> As for the pgf-grammars. I've also kinda changed my mind about how this is currently implemented. Ideally the application should load the compiled grammar files from a prioritized list of data-directory e.g. in the case of XDG compliant operating systems it would be something like:

Yes, default search paths to look for grammars is good. But it shouldn't load all PGFs in those places, only the ones that are listed in the Yaml.

**Comment by fredefox** [comment]
I was thinking it could do it on a lazy basis. Keep the ones it already read in-memory and a grammar is requested that is not in the internal map, then we look in those paths for a suitable grammar. Probably just by looking at the filepath like we do now.

**Comment by fredefox** [comment]
I'd like to work on this now, unless you have something which is a bigger priority.

**Comment by heatherleaf** [comment]
I think that's fine. You don't have to overwork though, here's how I think:

### Ajax

The grammars are loaded when `muste-ajax` is invoked, and are read from the Yaml config file. It's not necessary to be able to load grammars while the server is running.

### CLI

The CLI can be invoked as before, and the grammar is searched among the default grammar directories. The grammar can also be specified as a direct path to the file:
```
muste cli -G novo_modo/Prima.pgf -L PrimaSwe "hon är vis"
```

**Comment by fredefox** [comment]
Doesn't it make more sense to load the grammars based on whats in the DB? That's how we do it today.

**Comment by fredefox** [comment]
This is fully implemented in 2bcce666b74166893d139bd582abf0596676e2dc. Unfortunately f2c10e7 introduces a significant performance regression. I don't understand why... I think this error needs to be found before it makes sense to merge this in.

**Comment by fredefox** [comment]
Oh, and you can see how the config is supposed to look like here:

https://github.com/MUSTE-Project/MULLE/blob/2bcce666b74166893d139bd582abf0596676e2dc/muste-ajax/data/lessons.yaml

**Comment by fredefox** [comment]
It seems like only some lessons are slow to fetch. The `Exemplum`-based exercises seem to be doing ok... Really strange.

**Comment by fredefox** [comment]
I've merged this in and made a new release with our current changes.

**Comment by heatherleaf** [comment]
Thanks! I look into this over the weekend

**Comment by fredefox** [comment]
Don't know if you saw the comment I just made (and deleted again). I changed my mind about making a new "release" because I think that 4b7e515 needs to be reverted and we need to find another solution to #64.

**Comment by heatherleaf** [comment]
Reopening this, because I don't want the lesson config files to be loaded at compile time, as it seems to be now.

1. I compile everything
2. I run `muste-ajax --recreate-db`
3. I go to the web interface, select the Exemplum Eng->Swe lesson and click "god". I get the following menu:

    > en [] kung läser en bok
    > en [blå] kung läser en bok
    > en [dålig] kung läser en bok
    > en [kall] kung läser en bok
    > en [stor] kung läser en bok

    This is all correct so far. 

4. Now I edit `lessons.yaml`: https://github.com/MUSTE-Project/MULLE/blob/60838bca22a7341a5bea6e4e2eec3a77dbd06a16/muste-ajax/data/lessons.yaml#L92-L95
    I change `search-options:size` from 7 to 1

5. I stop the server and run `muste-ajax --recreate-db` again, go to the GUI, select the lesson and click "god"
6. I still get the same menu as before, but with the new reduced search option I shouldn't have got the first menu item ("en [] kung läser en bok"). (This can be seen by running `muste cli` with the same options)

So, the config `lessons.yaml` is apparently loaded at compile time, but I don't know where.

**Comment by fredefox** [comment]
> So, the config lessons.yaml is apparently loaded at compile time, but I don't know where.

Here:

https://github.com/MUSTE-Project/MULLE/blob/60838bca22a7341a5bea6e4e2eec3a77dbd06a16/muste-ajax/src/Muste/Web/Config.hs#L32-L33

**Comment by fredefox** [comment]
Oh wait. I misunderstood you. The config file I pointed you to is another one. If you look carefully at the output from `muste-ajax` on start-up it prints something like:

    lessons: /home/fredefox/git/MUSTE-Project/MULLE/.stack-work/install/x86_64-linux/15f0267747c5d7582a9f4feaa56e852d312d1790f584a6c9077902f187ce5508/8.4.3/share/x86_64-linux-ghc-8.4.3/muste-ajax-0.2.6.0/data/lessons.yaml

This is the path that the initial config file gets copied to. If you want to update the lessons of an already installed instance of `muste-ajax` then this is the place to do it.

I suspected you wanted a feature like this, so I designed it like this.
